### PR TITLE
Ai add pundit

### PR DESCRIPTION
**Pull request by aizatibraimova**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,12 +1,26 @@
...
+  include Pundit
 
+  before_action :authenticate_user!
   before_action :configure_permitted_parameters, if: :devise_controller?
 
+  after_action :verify_authorized, unless: :devise_controller?
+  after_action :verify_policy_scoped, only: :index, unless: :devise_controller?
+
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+
```
Comment (aborov) [inline app/controllers/application_controller.rb]: Great job setting up application-level "security by default" turning on Pundit and setting `rescue` and `after_action`

File: app/views/comments/show.html.erb [hunk 2]
```diff
@@ -14,5 +13,8 @@
   <%= @comment.body %>
 </p>
 
-<%= link_to 'Edit', edit_comment_path(@comment) %> |
+<% if policy(@comment).edit? %>
+  <%= link_to 'Edit', edit_comment_path(@comment) %> |
+<%= end %>
+
```
Comment (aborov) [inline app/views/comments/show.html.erb]: You might want to revise your comments view as currently I can see icons to edit and delete comments, even for the ones made by other users (the good thing when you click it it actually won't allow the action and you get the right alert, but you're not supposed to see those icons in the first place)

File: app/controllers/users_controller.rb
Comment (aborov) [inline app/controllers/users_controller.rb:1]: You might want to revise the policy on users (from the `users_controller` all the way down to the `show` view) as currently I can see photos from private users and their comments under photos

File: app/policies/photo_policy.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  attr_reader :user, :photo
+
+  def initialize(user, photo)
+    @user = user
+    @photo = photo
+  end
+
+  # Our policy is that a photo should only be seen by the owner or followers
+  #   of the owner, unless the owner is not private in which case anyone can
+  #   see it
```
Comment (aborov) [inline app/policies/photo_policy.rb]: You might want to specify cases when a user is trying to edit or delete their own photos. Currently they can't.

**Review by aborov — CHANGES_REQUESTED** [review]
It's a great start, but needs some adjustments. See my findings below.

File: app/controllers/application_controller.rb [hunk 4]
```diff
@@ -1,12 +1,26 @@
 class ApplicationController < ActionController::Base
   before_action :authenticate_user!
-
   before_action :configure_permitted_parameters, if: :devise_controller?
+  
+  include Pundit
+
```
Comment (ajmompr) [inline app/controllers/application_controller.rb]: great

File: app/controllers/application_controller.rb [hunk 5]
```diff
@@ -1,12 +1,26 @@
 class ApplicationController < ActionController::Base
   before_action :authenticate_user!
-
   before_action :configure_permitted_parameters, if: :devise_controller?
+  
+  include Pundit
+  
+  after_action :verify_authorized, unless: :devise_controller?
```
Comment (ajmompr) [inline app/controllers/application_controller.rb:7]: great!

File: app/controllers/application_controller.rb [hunk 6]
```diff
@@ -1,12 +1,26 @@
...
+  after_action :verify_policy_scoped, only: :index, unless: :devise_controller?
```
Comment (ajmompr) [inline app/controllers/application_controller.rb:8]: great!

File: app/controllers/comments_controller.rb
Comment (ajmompr) [inline app/controllers/comments_controller.rb:1]: Great start! You can refactor to be more DRY using before_action -> { authorize @comment || Comment },except: [:new, :create]. By calling this before action in your controller, you can remove each instance of authorize @comment. Same concept can be applied across your controllers

File: app/controllers/follow_requests_controller.rb [hunk 7]
```diff
@@ -1,9 +1,11 @@
 class FollowRequestsController < ApplicationController
   before_action :set_follow_request, only: %i[ show edit update destroy ]
+  before_action :authorize_follow_request, only: %i[accept destroy]
```
Comment (ajmompr) [inline app/controllers/follow_requests_controller.rb]: Refactor to be more DRY! replace with before_action { authorize @follow_request || FollowRequest } and see notes from your comments controller

File: app/controllers/follow_requests_controller.rb [hunk 8]
```diff
@@ -1,9 +1,11 @@
...
 
   # GET /follow_requests or /follow_requests.json
   def index
-    @follow_requests = FollowRequest.all
+    @follow_requests = policy_scope(FollowRequest)
+    authorize @follow_request
```
Comment (ajmompr) [inline app/controllers/follow_requests_controller.rb]: refactor to be more DRY

File: app/controllers/follow_requests_controller.rb
Comment (ajmompr) [inline app/controllers/follow_requests_controller.rb:1]: refactor to be more DRY

File: app/controllers/photos_controller.rb [hunk 9]
```diff
@@ -1,10 +1,10 @@
 class PhotosController < ApplicationController
   before_action :set_photo, only: %i[ show edit update destroy ]
-  before_action :ensure_current_user_is_owner, only: [:destroy, :update, :edit]
+  before_action :authorize_photo, except: %i[ index new create ]
```
Comment (ajmompr) [inline app/controllers/photos_controller.rb]: before_action { authorize @photo || Photo }

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -1,13 +1,62 @@
 class UsersController < ApplicationController
   before_action :set_user, only: %i[ show liked feed followers following discover ]
+  before_action :authorize_user, except: %i[ index new create ]
```
Comment (ajmompr) [inline app/controllers/users_controller.rb]: before_action { authorize @user || User }

File: app/controllers/users_controller.rb
Comment (ajmompr) [inline app/controllers/users_controller.rb:1]: refactor to be more DRY

File: app/policies/application_policy.rb
Comment (ajmompr) [inline app/policies/application_policy.rb:1]: good job using pundit generator

File: app/policies/comment_policy.rb [hunk 11]
```diff
@@ -0,0 +1,40 @@
+class CommentPolicy < ApplicationPolicy
+  attr_reader :user, :comment
+
+  def initialize(user, comment)
+    @user = user
+    @comment = comment
+  end
+
+  def show?
+    user == comment.author || !comment.author.private? || comment.author.followers.include?(user)
```
Comment (ajmompr) [inline app/policies/comment_policy.rb]: this is okay; however, we want to display all comments. If a photo is private then, we would not need this logic

File: app/policies/comment_policy.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+  def create?
+    !user.nil?
+  end
+
+  def update?
+    user == comment.author
+  end
+
+  class Scope < Scope
```
Comment (ajmompr) [inline app/policies/comment_policy.rb]: scope can just be inherited from application policy

File: app/policies/follow_request_policy.rb [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+  def new?
+    create?
+  end
+
+   def create?
+    !user.nil?
+  end
+
+   def update?
```
Comment (ajmompr) [inline app/policies/follow_request_policy.rb]: Who else can update? Remember, It takes two users to exchange follows

File: app/policies/follow_request_policy.rb [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    user == follow_request.sender || user == follow_request.recipient
+  end
+
+
+  def edit?
+    user == comment.author
+  end
+
+  def destroy?
+    user == follow_request.sender || follow_request.recipient
```
Comment (ajmompr) [inline app/policies/follow_request_policy.rb]: good

File: app/policies/follow_request_policy.rb [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    @follow_request = follow_request
+  end
+
+  def show?
+    user == follow_request.sender || user == follow_request.recipient
+  end
+
+
+  def edit?
+    user == comment.author
```
Comment (ajmompr) [inline app/policies/follow_request_policy.rb]: Why are we calling comment.author in follow request policy?

File: app/policies/photo_policy.rb (hunk 3 above)
Comment (ajmompr) [reply app/policies/photo_policy.rb]: added on lines 24-33

File: app/policies/photo_policy.rb [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+  def show?
+    user == photo.owner || !photo.owner.private? || photo.owner.followers.include?(user)
```
Comment (ajmompr) [inline app/policies/photo_policy.rb:10]: good

File: app/policies/user_policy.rb [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    @current_user = current_user
+    @user = user
+  end
+
+  def show?
+    true
+  end
+
+  def feed?
+    true
```
Comment (ajmompr) [inline app/policies/user_policy.rb:18]: Not quite. Should user == current_user

File: app/policies/user_policy.rb [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+  end
+
+  def edit?
+    update?
+  end
+
+  def update?
```
Comment (ajmompr) [inline app/policies/user_policy.rb:33]: Do we need to give users edit and update access based on the current version of the photogram app? Also, nitpick; if you are going to use inheritance, place update? before edit?

File: app/policies/user_policy.rb [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    user == current_user || current_user.admin?
+  end
+
+  def destroy?
```
Comment (ajmompr) [inline app/policies/user_policy.rb]: Do we need to give current_user the ability to destroy user record?

File: app/views/users/show.html.erb [hunk 20]
```diff
@@ -4,7 +4,7 @@
   </div>
 </div>
 
-<% if current_user == @user || !@user.private? || current_user.leaders.include?(@user) %>
+<% if policy(@user).show? %>
```
Comment (ajmompr) [inline app/views/users/show.html.erb]: Almost, but not quite. You need to add another policy method to ensure that the photo feed is displaying properly for both public and private users

**Review by ajmompr — CHANGES_REQUESTED** [review]
great start; see comments

File: app/controllers/follow_requests_controller.rb [hunk 21]
```diff
@@ -1,11 +1,10 @@
 class FollowRequestsController < ApplicationController
   before_action :set_follow_request, only: %i[ show edit update destroy ]
-  before_action :authorize_follow_request, only: %i[accept destroy]
+  before_action -> { authorize @follow_request || FollowRequest }, only: [:show, :edit, :update, :destroy]
```
Comment (aborov) [inline app/controllers/follow_requests_controller.rb:3]: Great job refactoring your controller to make your code more concise! Please double check your restrictions as it seems like a current user isn't authorized to accept pending requests.

File: app/controllers/comments_controller.rb [hunk 22]
```diff
@@ -1,6 +1,6 @@
 class CommentsController < ApplicationController
   before_action :set_comment, only: %i[ show edit update destroy ]
-  before_action :authorize_comment, except: %i[ index new create ]
+  before_action -> { authorize @comment || Comment }, except: %i[new create]
```
Comment (aborov) [inline app/controllers/comments_controller.rb:3]: Nice work refactoring comments! It is indeed more consice now.

File: app/controllers/users_controller.rb [hunk 23]
```diff
@@ -1,9 +1,6 @@
 class UsersController < ApplicationController
   before_action :set_user, only: %i[ show liked feed followers following discover ]
-  before_action :authorize_user, except: %i[ index new create ]
-
-  # after_action :verify_authorized, except: :index
-  # after_action :verify_policy_scoped, only: :index
+  before_action -> { authorize @user || User }
```
Comment (aborov) [inline app/controllers/users_controller.rb:3]: Awesome work using concise authorization for users! Now you can actually take out almost all of your actions in this controller!

File: app/policies/comment_policy.rb [hunk 24]
```diff
@@ -7,34 +7,22 @@ def initialize(user, comment)
...
+    true
+  end
+  
+  def create?
+    true
   end
 
   def edit?
     update?
   end
```
Comment (aborov) [inline app/policies/comment_policy.rb]: you should be good to take many of these actions out

File: app/policies/follow_request_policy.rb [hunk 25]
```diff
@@ -7,35 +7,18 @@ def initialize(user, follow_request)
...
-
-  def new?
-    create?
+    user == follow_request.recipient
   end
 
-   def create?
-    !user.nil?
+  def create?
+    true
```
Comment (aborov) [inline app/policies/follow_request_policy.rb:14]: good!

File: app/policies/follow_request_policy.rb [hunk 26]
```diff
@@ -7,35 +7,18 @@ def initialize(user, follow_request)
...
   end
 
-  class Scope < Scope
-    def resolve
-      scope.all.select do |follow_request|
-        user == follow_request.sender || user == follow_request.recipient
-      end
-    end
+  def destroy?
+    user == follow_request.sender || user == follow_request.recipient
```
Comment (aborov) [inline app/policies/follow_request_policy.rb:22]: also you can just say:
```ruby
def destroy?
  update?
end
```

File: app/policies/follow_request_policy.rb [hunk 27]
```diff
@@ -7,35 +7,18 @@ def initialize(user, follow_request)
...
-    user == comment.author
-  end
-
-  def destroy?
-    user == follow_request.sender || follow_request.recipient
-  end
-
-  def new?
-    create?
+    user == follow_request.recipient
```
Comment (aborov) [inline app/policies/follow_request_policy.rb:10]: I think you don't need it

File: app/policies/photo_policy.rb [hunk 28]
```diff
@@ -17,27 +14,15 @@ def create?
     true
   end
 
-  def new?
-    create?
-  end
-
   def update?
     user == photo.owner
   end
```
Comment (aborov) [inline app/policies/photo_policy.rb]: You might want to revise it: currently, when a user is trying to update their own photo, they are not authorized for that

File: app/policies/photo_policy.rb [hunk 29]
```diff
@@ -17,27 +14,15 @@ def create?
...
-  def edit?
-    update?
-  end
+  
+    def edit?
+      update?
+    end
 
   def destroy?
     user == photo.owner
```
Comment (aborov) [inline app/policies/photo_policy.rb:26]: this can be simplified to 
```ruby
def destroy?
  update?
end
```

File: app/policies/user_policy.rb
Comment (aborov) [inline app/policies/user_policy.rb:1]: `discover` currently doesn't work, so you might want to find out what's broken.
Also, you currently can't access profile pages for any private users at all.

File: app/views/users/show.html.erb [hunk 30]
```diff
@@ -4,7 +4,7 @@
   </div>
 </div>
 
-<% if policy(@user).show? %>
+<% if policy(@user).show_photos? %>
```
Comment (aborov) [inline app/views/users/show.html.erb:7]: awesome!

**Review by aborov — CHANGES_REQUESTED** [review]
that's a great progress! however, there are still some spots that call for revision. please see my comments below.

File: app/policies/user_policy.rb [hunk 31]
```diff
@@ -0,0 +1,44 @@
...
+  def feed?
+    true
+  end
+
+  def update?
+    user == current_user
+  end
+
+  def edit?
+    update?
```
Comment (aborov) [inline app/policies/user_policy.rb:38]: great use of inheritance!

File: app/views/comments/_comment.html.erb [hunk 32]
```diff
@@ -9,14 +9,16 @@
       </h5>
       <p><%= comment.body %></p>
     </div>
-    <div>
-      <%= link_to edit_comment_path(comment), class: "btn btn-link btn-sm text-muted" do %>
-        <i class="fas fa-edit fa-fw"></i>
-      <% end %>
+    <% if comment.author== current_user %>
```
Comment (aborov) [inline app/views/comments/_comment.html.erb:12]: nice job hiding editing buttons from users that don't own this comment!

File: app/views/users/_user.html.erb [hunk 33]
```diff
@@ -98,16 +100,18 @@
     <div class="row mb-4">
       <div class="col">
         <% follow_request = current_user.sent_follow_requests.find_by(recipient: @user) %>
-        <% if follow_request %>
-          <%= link_to follow_request, data: { turbo_method: :delete }, class: "btn btn-outline-secondary btn-block" do %>
-            <% if follow_request.pending? %>
-              Un-request
-            <% else %>
-              Un-follow
+        <% if @user != current_user %>
```
Comment (aborov) [inline app/views/users/_user.html.erb:103]: nice way to hide pending requests for users who are looking at the profile that's not their own!

**Review by aborov — APPROVED** [review]
great job, Aizat!

File: app/controllers/follow_requests_controller.rb [hunk 34]
```diff
@@ -1,9 +1,10 @@
 class FollowRequestsController < ApplicationController
   before_action :set_follow_request, only: %i[ show edit update destroy ]
+  before_action -> { authorize @follow_request || FollowRequest }, only: [:show, :edit, :update, :destroy]
 
   # GET /follow_requests or /follow_requests.json
   def index
-    @follow_requests = FollowRequest.all
+    @follow_requests = policy_scope(FollowRequest)
```
Comment (ajmompr) [inline app/controllers/follow_requests_controller.rb:7]: redundant if you are calling before_action! But okay

File: app/controllers/photos_controller.rb [hunk 35]
```diff
@@ -14,6 +14,7 @@ def show
   # GET /photos/new
   def new
     @photo = Photo.new
+    authorize @photo
```
Comment (ajmompr) [inline app/controllers/photos_controller.rb:17]: redundant if calling before_action but okay

**Review by ajmompr — APPROVED** [review]
A lot of the work that you are doing is redundant because you have added the before_action { authorize ...} in your controllers. Be sure to stay consistent when adding your authorizations

**Pull request merged**